src/metabase/query_processor/pivot/postprocess.clj

```diff
@@ -339,45 +341,44 @@
+        row-combos              (mapv vec (apply math.combo/cartesian-product (mapv row-values pivot-rows)))
+        col-combos              (mapv vec (apply math.combo/cartesian-product (mapv column-values pivot-cols)))
+        col-combos              (vec (sort-column-combos config col-combos))
         row-totals?             (and row-totals? (boolean (seq pivot-cols)))
         column-headers          (build-column-headers config col-combos col-formatters)
-        headers                 (or (seq (build-headers column-headers config))
-                                    [(concat
-                                      (map #(get column-titles %) pivot-rows)
-                                      (map #(get column-titles %) pivot-measures))])]
-    (concat
-     headers
-     (filter seq
-             (apply concat
-                    (let [sections-rows
-                          (for [section-row-combos ((get sort-fns (first pivot-rows) identity) (sort-by ffirst (vals (group-by first row-combos))))]
-                            (concat
-                             (remove nil?
-                                     (for [row-combo section-row-combos]
-                                       (build-row row-combo col-combos pivot-measures data totals row-totals? ordered-formatters row-formatters)))))]
-                      (mapv
-                       (fn [section-rows]
-                         (->>
-                          section-rows
-                          (sort-pivot-subsections config)
-                          ;; section rows are either enriched with column-totals rows or left as is
-                          ((fn [rows]
-                             (if (and col-totals? (> (count pivot-rows) 1))
-                               (append-totals-to-subsections pivot rows col-combos ordered-formatters)
-                               rows)))
-                          ;; then, we apply the row-formatters to the pivot-rows portion of each row,
-                          ;; filtering out any rows that begin with "Totals ..."
-                          (mapv
-                           (fn [row]
-                             (let [[row-part vals-part] (split-at (count pivot-rows) row)]
-                               (if (or
-                                    (not (seq row-part))
-                                    (str/starts-with? (first row-part) "Totals"))
-                                 row
-                                 (vec (concat (map fmt row-formatters row-part) vals-part))))))))
-                       sections-rows))))
-     (when col-totals?
-       [(build-grand-totals config col-combos totals row-totals? ordered-formatters)]))))
+        headers                 (or (not-empty (build-headers column-headers config))
+                                    [(mapv #(get column-titles %) (into (vec pivot-rows) pivot-measures))])]
+    (-> headers
```

Comment:
Everything above is assuredly a vector now. I suspect we should add a `(vec headers)` here as if we do something silly and make one of the above a seq the `not-empty` check will propagate that and our ordering of everything here will be lost.

Reply:
I know you are pointing out the evils of concat, but does it have it's purpose here if we did something like

`(into [] (concat headers section-rows grand-totals)` ? My thinking is that the legibility is much higher and paying the cost of one concat is not that high as part of the final step to bring it all together?

Reply:
You are right that `(into [] (concat headers section-rows grand-totals)` very likely wouldn't cause much of a perf drop. But I still think that not using concat ever is better from hygiene standpoint, it is a function that can go wrong very easily, not only performance-wise but also functionally (https://stuartsierra.com/2015/04/26/clojure-donts-concat).

I think there is a point of introducing a custom `metabase.util.performance/concat` and use it instead if one wants the legibility of concat.